### ISSUE TITLE
Support making any container configuration key into a column in list

### ISSUE THREAD

**Issue by cwre** [lxc/lxd#2032]
# Feature Request

Add a comment, image, or some form of identifier to `lxc list` please. It would be incredibly helpful!

Example of what I mean:
```
+-------------------+---------+----------------------+------+------------+-----------+--------------------------------+
|       NAME        |  STATE  |         IPV4         | IPV6 |    TYPE    | SNAPSHOTS |           COMMENT              |
+-------------------+---------+----------------------+------+------------+-----------+--------------------------------+
| venom             | STOPPED |                      |      | PERSISTENT | 0         |       16.04 Test Container     | 
+-------------------+---------+----------------------+------+------------+-----------+--------------------------------+
```

**Comment by tych0** [comment]
On Wed, May 18, 2016 at 10:33:01AM -0700, Alex Roberts wrote:
> # Feature Request
> 
> Add a comment, image, or some form of identifier to `lxc list` please. It would be incredibly helpful!
> 
> Example of what I mean:
> ```
> +-------------------+---------+----------------------+------+------------+-----------+-----------+
> |       NAME        |  STATE  |         IPV4         | IPV6 |    TYPE    | SNAPSHOTS |COMMENT |
> +-------------------+---------+----------------------+------+------------+-----------+-----------+
> | venom          | STOPPED |                      |      | PERSISTENT | 0         |16.04 Test Container       |
> +-------------------+---------+----------------------+------+------------+-----------+-----------+
> ```

as I said on IRC, you can set user.* keys,

lxc config set venom user.description "16.04 Test Container"

**Comment by cwre** [comment]
Sadly, this isn't as useful as I would like it.

I can list by the description manually, but I would just like to be able to visually grep for the container I need after running `lxc list` rather than having to run `lxc list user.description="Foo"`.

**Comment by tych0** [comment]
On Wed, May 18, 2016 at 11:42:51AM -0700, Alex Roberts wrote:
> Sadly, this isn't as useful as I would like it.
> 
> I can list by the description manually, but I would just like to be able to visually grep for the container I need after running `lxc list` rather than having to run `lxc list user.description="Foo"`.

Sounds like maybe you really want the ability to add configuration
keys as a --column to lxc list then?

**Comment by cwre** [comment]
Yes, I suppose that would be more accurate.

**Comment by ctrlrsf** [comment]
Any thoughts on how the --columns flag should support this? Would comma separated strings following the column characters for the table format suffice?

For example:
```
lxc list --columns ns46tS,user.key1,user.key2
```

**Comment by cwre** [comment]
I'm not sure I understand...

So if I were to do  `lxc list --columns user.description` it would append another column to the list command with the keys for user.description for each container?

Yes, that would be fine.

**Comment by ctrlrsf** [comment]
OK, just trying to understand how this would be implemented since right now --columns takes a series of characters to specify a column layout format. Specifically, how it would distinguish a series of keys vs a series of characters for the layout format.

**Comment by cwre** [comment]
Ahh, good point.

**Comment by stgraber** [comment]
Yeah, I think supporting a comma separate list of config keys would do the trick. We may even support something like:

lxc list --columns ns46tS,user.description:Description,user.owner:Owner

As a way to prettify the output. This would like the following columns:
 - Name
 - Status
 - IPv4
 - IPv6 
 - Type
 - Snapshots
 - Description
 - Owner

**Comment by sean-jc** [comment]
Regarding the output, having a way to specify the max width of the output of a column would be helpful.  Maybe something like `user.description:Description(20)` or `user.description:Description:20` to specify the max characters.  The max characters value could either 1) default to the width of the column header, allowing 0 or -1 to specify unlimited, or 2) default to unlimited, with a special value (`n` for natural?) setting it to the header width.

**Comment by stgraber** [comment]
Yeah, I'd be fine with supporting truncation, ```user.description:Description:20``` would be my preference as it won't mess with your shell as much as brackets do.

As for the default, I'd go with unlimited being the default (so column width matches largest value's length) but wouldn't mind having a special value for truncating to the width of the column name.

So maybe:
 - user.description => Column name is "user.description", no width limit
 - user.description:Description => Column name is "Description", no width limit
 - user.description:Description:10 => Column name is "Description", width limit set to 20
 - user.description:Description:-1 => Column name is "Description", width limit is the length of the column name

Basically making "0" be unlimited (the default) and "-1" be the length of the column header.

**Comment by sean-jc** [comment]
I've got the functional portion of this coded up, I'll work on documentation and testing and hopefully get a PR submitted in the next day or two.

**Comment by stgraber** [comment]
This got implemented.
